Cody Agent: reuse TypeScript logic between JetBrains and VS Code

olafurpg opened the pull request:
Previously, the JetBrains plugin had to manually translate all the
prompt logic from TypeScript from Java. This commit fixes the problem
by introducing a Cody "agent" that runs as a JSON-RPC server and
encapsulates all the prompt logic. The JetBrains plugin launches the
Cody agent server during startup and communicates via the agent protocol
to accomplish tasks like listing recipes and executing recipes.

This commit only updates the JetBrains plugin to use the "chat-question"
recipe. Remaining functionality like other recipes and completions will
need to migrate to the agent in follow-up PRs.

Based on https://github.com/sourcegraph/sourcegraph/pull/52800 by @SuperAuguste 

More details in the RFC  https://docs.google.com/document/d/1KEJEd_Xx8STbJeGyOo86dkh3QRHQMs4CPw4XRAm7o7U/edit#heading=h.trqab8y0kufp

## Test plan

Green CI.

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

dominiccooney commented:
Looks like there's some typescript work required.

What's motivating the changes like Foo | undefined => Foo | null? Would it be better to tread lightly? Foo | undefined is more succinct TypeScript because the implicit return is undefined, not null.

What motivates making this more async? Things using the agent want to use the network, hence be async most everywhere? This might be inconvenient for vscode APIs which provide/consume text synchronously; I also wonder how brittle our tests will be if the timing of updates changes... @philipp-spiess do Playwright locators have implicit "wait for ..." behavior in them?

olafurpg commented on client/cody-shared/src/chat/client.ts:
```diff
@@ -52,10 +56,11 @@ export async function createClient({
     const fullConfig = { debugEnable: false, ...config }
 
-    const completionsClient = new SourcegraphBrowserCompletionsClient(fullConfig)
+    const completionsClient = createCompletionsClient(fullConfig)
```
This diff is necessary to avoid a runtime Node.js crash from using browser APIs

olafurpg reviewed (commented):
@dominiccooney me and @SuperAuguste paired on this and were able to remove all the async/null related diffs in cody-shared. This more or less eliminates the risk that merging this PR will introduce regressions in the vscode/web clients.

We can discuss separately how to design a new JSON-RPC API for non-stop/fixup/tools/plugins. I'd love to merge this PR asap to unblock further progress on the JetBrains plugin (also Emacs/NeoVim)

philipp-spiess commented:
>  I also wonder how brittle our tests will be if the timing of updates changes... @philipp-spiess do Playwright locators have implicit "wait for ..." behavior in them?

Yeah pretty much all of the locators have that built-in, see: https://playwright.dev/docs/actionability

vdavid commented on client/cody-agent/package.json:
```diff
@@ -0,0 +1,29 @@
+    "make-all-exes": "pnpm run build && pkg -t node16-linux-arm64,node16-linux-x64,node16-macos-arm64,node16-macos-x64,node16-win-x64 dist/agent.js --out-path ${AGENT_EXE_DEST:-dist}",
+    "lint": "pnpm run lint:js",
+    "lint:js": "eslint --cache '**/*.[tj]s?(x)'",
+    "test": "npm run build && jest"
```
Nit: should this be pnpm?

vdavid reviewed (commented):
I went through the code a few times.
It makes sense to me.
I saw the bunch of TODOs, and I think it's fine to merge this as-is, and continue improving/completing it from this point on.

Notes from the JetBrains POV:
- I've tried to run the branch, including the agent, but the docs are not there yet—even if I built and ran /client/cody-agent/dist/index.js with Node, the plugin couldn't connect to it. I'm probably missing a piece there.
- Regardless of the answer, the agent doesn't run automatically in this version, which breaks the chat in the plugin. I want to periodically release and dogfood the plugin, so I want to avoid a broken version of it on `main`. For this reason, I suggest to exclude the parts of the code from this PR that hook up the plugin code with the agent. I'm fine with merging the rest, and actually, it'd be great to have it in.

olafurpg commented:
@vdavid I'm working on resolving this comment here https://github.com/sourcegraph/sourcegraph/pull/53370#discussion_r1229906435 so that it's possible to run the agent locally on any computer. I'm updating the Gradle build to build standalone binaries for the agent. 

Agreed that we should not merge this PR with regressions in the JetBrains plugin, it needs to work with full feature parity as before this PR.

dominiccooney commented on client/cody-agent/src/index.test.ts:
```diff
@@ -0,0 +1,100 @@
+import { MessageHandler } from './rpc'
+
+export class TestClient extends MessageHandler {
+    constructor() {
```
The implicit constructor does what you need. I think there should be a lint complaining about this...

dominiccooney commented on client/cody-agent/src/rpc.ts:
```diff
@@ -0,0 +1,246 @@
+                    }
+
+                    switch (headerName) {
+                        case 'Content-Length':
```
What protocol is this; can I see the documentation for this header?

SuperAuguste replied:
Content-Length describes the length of the body in bytes, see [these docs which we based the protocol on](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#headerPart). I'll add some comments to explain this. :)

dominiccooney commented on client/cody-agent/src/rpc.ts:
```diff
@@ -0,0 +1,246 @@
+                            break
+                    }
+
+                    startIndex = endIndex + 2
```
I gave the same feedback on @SuperAuguste 's PR but the code would be more readable if you did something like '\r\n'.length so it is really clear where this magic number comes from.

I'm going to stop reviewing now; maybe run through the feedback on the earlier PR and let me know when you're ready to take another look.

dominiccooney reviewed (commented):
The idea of plumbing from VSCode to cody-shared, to cody-agent, and back again is pretty unappetizing. All the string serialization and parsing implied also makes me wonder how responsive we can make this. But @olafurpg promises me that there's a glorious future on the other side of this so let's see if we can make that happen.

Some things I would be interested in having the agent give us:
- A nicely container for background threads so we block VScode less.
- A layer of capability advertisement and negotiation so supporting different product configurations to match the capability of employed LLMs happens "for no extra cost."
- A place for plugging in and orchestrating third party components so we can bootstrap some network effects.
- Something that can support multiple concurrent clients so starting a new editor is faster.

Some things which would suggest we reevaluate:
- Developers are complaining about resource consumption.
- Update rates suffer because our packages are bigger.
- We can't do really interactive features because round trips or streaming is too slow.
- The multiple editors as clients don't materialize.
- Toil maintaining the plumbing through these layers eats too much into time working on new features.
- The versioning, capability, editor matrix becomes sclerotic and there's confusion about what features work where.

I think my earlier feedback on @SuperAuguste 's version of this might have been lost, so let me leave this with you and please ping when it is ready for another look.

SuperAuguste commented on client/cody-agent/src/rpc.ts:
```diff
@@ -0,0 +1,246 @@
+                        // This state is irrecoverable because the stream is polluted
+                        // Also what is the client doing 😭
+                        this.contentLengthRemaining = newContentLength
+                        assert(this.contentLengthRemaining !== null)
```
Yep, as I mentioned there this is left-over code from a while back, I'll remove it.

olafurpg commented:
Status update on this PR:

- The JetBrains plugin side needs more work
  - The agent binaries are available via `./gradlew :runIde` but not when the plugin is installed via zip file
  - Error handling and debugging 
- @SuperAuguste is looking into addressing the review on the TypeScript side
- I'm OOO tomorrow but will resume work on this PR next week

valerybugakov commented on client/cody-agent/src/protocol.ts:
```diff
@@ -0,0 +1,108 @@
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+export type Notifications = {
+    // Client -> Server
+    initialized: [void]
```
Do we have to use `void` here and in other places in the `cody-agent` package? Ideally, we should rely only on `undefined` and `null`.

valerybugakov commented on client/cody-agent/src/rpc.ts:
```diff
@@ -0,0 +1,252 @@
+
+import { Notifications, Requests } from './protocol'
+
+export type RequestMethod = keyof Requests
+export type NotificationMethod = keyof Notifications
+
+export type Method = RequestMethod | keyof Notifications
+export type ParamsOf<K extends Method> = (Requests & Notifications)[K][0]
+export type ResultOf<K extends RequestMethod> = Requests[K][1]
```
It would be great to add comments here explaining the expected structure of `Requests` and `Notifications` and how they play with `ParamsOf` and `ResultOf` to make it easier to read.

valerybugakov reviewed (commented):
Interesting read. I left a couple of comments for the TS diff. Echoing Dominic's message, it would be helpful to expand the README of the new package with more details.

olafurpg commented:
Update since last changes:

- The Gradle build now automatically builds agent binaries and embeds them with the IntelliJ plugin via Gradle 
- Did a lazy rebase on master, which added 2.5k diff on pnpm-lock that needs to get fixed

vdavid commented:
@olafurpg is there a follow-up PR for this? If so, could you link the two?

olafurpg commented on client/cody-agent/jest.config.js:
```diff
@@ -0,0 +1,11 @@
+/** @type {import('@jest/types').Config.InitialOptions} */
+module.exports = {
+  ...config,
+  displayName: 'cody-shared',
```
Fixed.

olafurpg commented on client/cody-agent/README.md:
```diff
@@ -0,0 +1,3 @@
+# Cody agent
```
Good idea. Done.

olafurpg reviewed (commented):
Plan for tomorrow:

- Self-review the entire diff. I've been busy cleaning up many small things but haven't done a full review in awhile.
- Test the actual plugin more thoroughly.
- Fix error handling in the JetBrains side, but I may be tempted to postpone this for a separate PR.
- If nothing big comes up, will tag for review and hopefully we can merge this PR this week. I'll find somebody else to take over this PR if we can't make it before Friday when I go on vacation.

olafurpg commented on client/jetbrains/build.gradle.kts:
```diff
@@ -76,15 +79,15 @@ tasks {
-                    }
-                    subList(indexOf(start) + 1, indexOf(end))
-                }.joinToString("\n").run { markdownToHTML(this) }
+            projectDir.resolve("README.md").readText().lines().run {
```
This diff happened because I reformatted the file to fix indentation in other lines that I was working on. It would be nice to configure something like ktfmt to format the build script but it's not a priority for now. Maybe if/when we add Kotlin to the codebase.

vdavid reviewed (approved):
Stamping to unblock. I haven't gone through _all_ files in the final version, I'm at about 40%, but everything looked good so far, plus I've gone through earlier versions of the code several times. If I find something, we should be okay to fix it in a follow-up PR
